File: 22. Generate Parentheses/Solution.py
from typing import List

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def dfs(left, right, s):
            if len(s) == n * 2:
                res.append(s)
                return 

            if left < n:
                dfs(left + 1, right, s + '(')

            if right < left:
                dfs(left, right + 1, s + ')')

        res = []
        dfs(0, 0, '')
        return res

# Solicitar entrada do usuário
n = int(input("Input: 3"))

# Criar uma instância da classe Solution
solution = Solution()

# Obter a lista de strings de parênteses
output = solution.generateParenthesis(n)

# Exibir a saída formatada
print("Output:", output)

File: 22. Generate Parentheses/Solution.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Solution {

    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        dfs(0, 0, n, "", result);
        return result;
    }

    private void dfs(int left, int right, int n, String s, List<String> result) {
        if (s.length() == n * 2) {
            result.add(s);
            return;
        }

        if (left < n) {
            dfs(left + 1, right, n, s + "(", result);
        }

        if (right < left) {
            dfs(left, right + 1, n, s + ")", result);
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = 3;
        System.out.println("Input: " + n);

        Solution solution = new Solution();
        List<String> output = solution.generateParenthesis(n);

        System.out.println("Output: " + output);

        n = 1;
        output = solution.generateParenthesis(n);
        System.out.println("Output: " + output);
    }
}
